tools/SourceKit/tools/sourcekitd/bin/XPC/Client/sourcekitd.cpp

```diff
@@ -172,6 +183,9 @@ void sourcekitd_send_request(sourcekitd_object_t req,
 
   xpc_object_t msg = xpc_dictionary_create(nullptr, nullptr,  0);
   xpc_dictionary_set_value(msg, "msg", contents);
+  if (request_handle) {
+    xpc_dictionary_set_value(msg, "cancelToken", request_handle);
```

Comment:
* `xpc_dictionary_set_value` expects an `xpc_object_t`. I think you want `xpc_dictionary_set_uint64`
* This string constant could go in `Internal-XPC.h` next to `KeyInternalMsg`.
* Nit: we've used underscores instead of camelCase for these names.

Reply:
> * `xpc_dictionary_set_value` expects an `xpc_object_t`. I think you want `xpc_dictionary_set_uint64`

I thought because `xpc_object_t`’s definition is `typedef void * xpc_object_t`, `xpc_object_t` would just be an opaque pointer (which is what we are using for the cancellation token as well). Or am I making a wrong assumption here?

> * This string constant could go in `Internal-XPC.h` next to `KeyInternalMsg`.

I guess the `"msg"` string should also go into `Internal-XPC.h` as well in that case, right?

> * Nit: we've used underscores instead of camelCase for these names.

👍

Reply:
> I thought because xpc_object_t’s definition is typedef void * xpc_object_t, xpc_object_t would just be an opaque pointer (which is what we are using for the cancellation token as well). Or am I making a wrong assumption here?

`xpc_object_t` is `void *` because it's an opaque type in C, but it needs to be a real object created using one of the APIs like `xpc_dictionary_create()`.  Passing in an arbitrary pointer will crash.

We should probably add at least one test for this that uses `sourcekitd-test` so that we can catch this kind of issue.

Reply:
Ah, that makes sense. I also added cancellation support to `sourcekitd-test` now so we should be exercising that code path if `sourcekitd-test` is built with XPC.